File: R/plotting.R
library(tidyverse)

# Reusable ggplot2 helpers for synthetic Biochemistry teaching datasets.

require_columns <- function(df, columns) {
  missing <- setdiff(columns, names(df))
  if (length(missing) > 0) {
    stop(paste("Missing required columns:", paste(missing, collapse = ", ")))
  }
}

plot_group_means <- function(summary_df, group_col, mean_col, sem_col, title) {
  require_columns(summary_df, c(group_col, mean_col, sem_col))

  ggplot(summary_df, aes(x = .data[[group_col]], y = .data[[mean_col]])) +
    geom_col() +
    geom_errorbar(aes(
      ymin = .data[[mean_col]] - .data[[sem_col]],
      ymax = .data[[mean_col]] + .data[[sem_col]]
    ), width = 0.2) +
    labs(title = title)
}

plot_dose_response <- function(summary_df) {
  require_columns(summary_df, c("drug_name", "concentration_uM", "mean_viability", "sem_viability"))

  summary_df %>%
    mutate(plot_concentration = if_else(concentration_uM == 0, 0.001, concentration_uM)) %>%
    ggplot(aes(x = plot_concentration, y = mean_viability, color = drug_name)) +
    geom_line() +
    geom_point() +
    geom_errorbar(aes(ymin = mean_viability - sem_viability,
                      ymax = mean_viability + sem_viability), width = 0.05) +
    scale_x_log10() +
    labs(title = "Dose-Response Visualization",
         x = "Concentration (uM, log scale; 0 plotted as 0.001)",
         y = "Mean Cell Viability (%)")
}

plot_heatmap <- function(expr_df) {
  if (nrow(expr_df) == 0 || ncol(expr_df) == 0) {
    stop("expr_df must contain genes as rows and samples as columns")
  }

  expr_z <- t(scale(t(as.matrix(expr_df))))
  heatmap_df <- as.data.frame(expr_z) %>%
    mutate(gene = rownames(expr_z)) %>%
    pivot_longer(-gene, names_to = "sample", values_to = "z_score")

  ggplot(heatmap_df, aes(x = sample, y = gene, fill = z_score)) +
    geom_tile() +
    labs(title = "Gene Expression Heatmap")
}

make_pca_sample_scores <- function(expr_df) {
  if (nrow(expr_df) == 0 || ncol(expr_df) < 2) {
    stop("expr_df must contain genes as rows and at least two sample columns")
  }

  pca <- prcomp(t(expr_df), scale. = TRUE)
  as.data.frame(pca$x[, 1:2]) %>%
    mutate(sample = rownames(pca$x)) %>%
    mutate(group = str_extract(sample, "control|treat_low|treat_high|treatment"))
}

plot_pca_sample_map <- function(pca_df) {
  require_columns(pca_df, c("sample", "PC1", "PC2"))

  ggplot(pca_df, aes(x = PC1, y = PC2, color = group, label = sample)) +
    geom_point(size = 3) +
    geom_text(vjust = -0.8) +
    labs(title = "Toy PCA Sample Map")
}

File: src/plotting.py
"""Reusable Plotly plotting helpers for synthetic teaching datasets."""

import pandas as pd
import numpy as np
import plotly.express as px


def _require_columns(df: pd.DataFrame, columns) -> None:
    """Raise a clear error if expected columns are missing."""
    missing = [col for col in columns if col not in df.columns]
    if missing:
        raise ValueError(f"Missing required columns: {missing}")


def plot_group_means(summary_df: pd.DataFrame, group_col: str, mean_col: str, sem_col: str, title: str):
    """Create a bar chart with SEM error bars."""
    _require_columns(summary_df, [group_col, mean_col, sem_col])
    return px.bar(summary_df, x=group_col, y=mean_col, error_y=sem_col, title=title)


def plot_dose_response(summary_df: pd.DataFrame):
    """Create an interactive dose-response curve."""
    _require_columns(
        summary_df,
        ["drug_name", "concentration_uM", "mean_viability", "sem_viability", "sd_viability", "n"],
    )
    plot_df = summary_df.copy()
    plot_df["plot_concentration"] = plot_df["concentration_uM"].replace(0, 0.001)
    fig = px.line(
        plot_df,
        x="plot_concentration",
        y="mean_viability",
        color="drug_name",
        markers=True,
        error_y="sem_viability",
        hover_data=["concentration_uM", "sd_viability", "n"],
        title="Dose-Response Visualization",
    )
    fig.update_xaxes(type="log", title="Concentration (uM, log scale; 0 plotted as 0.001)")
    fig.update_yaxes(title="Mean Cell Viability (%)")
    return fig


def plot_fold_change(summary_df: pd.DataFrame):
    """Create a fold-change bar chart."""
    _require_columns(summary_df, ["gene", "log2_fold_change"])
    return px.bar(
        summary_df.sort_values("log2_fold_change"),
        x="gene",
        y="log2_fold_change",
        title="Toy Gene Expression: log2 Fold Change",
    )


def plot_heatmap(expr_df: pd.DataFrame):
    """Create a gene-wise z-score heatmap from a gene x sample expression matrix."""
    if expr_df.empty:
        raise ValueError("expr_df must contain at least one gene and one sample.")
    expr_z = expr_df.sub(expr_df.mean(axis=1), axis=0).div(expr_df.std(axis=1), axis=0)
    return px.imshow(expr_z, aspect="auto", title="Gene Expression Heatmap")


def make_pca_sample_scores(expr_df: pd.DataFrame) -> pd.DataFrame:
    """Create simple PCA-style sample scores from a gene x sample matrix.

    This uses NumPy SVD so students do not need an additional PCA package.
    """
    if expr_df.empty or expr_df.shape[1] < 2:
        raise ValueError("expr_df must include genes as rows and at least two sample columns.")
    sample_by_gene = expr_df.T
    scaled = (sample_by_gene - sample_by_gene.mean(axis=0)) / sample_by_gene.std(axis=0)
    scaled = scaled.replace([np.inf, -np.inf], np.nan).fillna(0)
    u, s, _ = np.linalg.svd(scaled, full_matrices=False)
    return pd.DataFrame({
        "sample": sample_by_gene.index,
        "PC1": u[:, 0] * s[0],
        "PC2": u[:, 1] * s[1] if len(s) > 1 else 0,
    })


def plot_pca_sample_map(pca_df: pd.DataFrame):
    """Create a PCA-style sample map from `make_pca_sample_scores()` output."""
    _require_columns(pca_df, ["sample", "PC1", "PC2"])
    plot_df = pca_df.copy()
    if "group" not in plot_df.columns:
        plot_df["group"] = plot_df["sample"].str.extract(r"(control|treat_low|treat_high|treatment)")
    return px.scatter(
        plot_df,
        x="PC1",
        y="PC2",
        color="group",
        text="sample",
        title="Toy PCA-style Sample Map",
    )
